src/jit/codegenxarch.cpp

```diff
@@ -1531,31 +1529,94 @@ CodeGen::genStructReturn(GenTreePtr treeNode)
 
         ReturnTypeDesc retTypeDesc;
         retTypeDesc.Initialize(compiler, varDsc->lvVerTypeInfo.GetClassHandle());
-        assert(retTypeDesc.GetReturnRegCount() == CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_RETURN_IN_REGISTERS);
-
-        var_types type0 = retTypeDesc.GetReturnRegType(0);
-        var_types type1 = retTypeDesc.GetReturnRegType(1);
+        unsigned regCount = retTypeDesc.GetReturnRegCount();
+        assert(regCount == MAX_RET_REG_COUNT);
 
-        regNumber reg0 = retTypeDesc.GetABIReturnReg(0);
-        regNumber reg1 = retTypeDesc.GetABIReturnReg(1);
-        assert(reg0 != REG_NA && reg1 != REG_NA);
-
-        // Move the values into the return registers     
-        getEmitter()->emitIns_R_S(ins_Load(type0), emitTypeSize(type0), reg0, lclVar->gtLclNum, 0);
-        getEmitter()->emitIns_R_S(ins_Load(type1), emitTypeSize(type1), reg1, lclVar->gtLclNum, 8);
+        // Move the values into ABI return registers     
+        int offset = 0;
+        for (unsigned i = 0; i < regCount; ++i)
+        {
+            var_types type = retTypeDesc.GetReturnRegType(i);
+            regNumber reg = retTypeDesc.GetABIReturnReg(i);
+            getEmitter()->emitIns_R_S(ins_Load(type), emitTypeSize(type), reg, lclVar->gtLclNum, offset);
+            offset += genTypeSize(type);
+        }
     }
     else
     {
-        // Assumption: multi-reg return value of a GT_CALL node is never spilled.
-        // TODO-BUG: support for multi-reg call nodes.
+        assert(op1->IsMultiRegCall());
+        genConsumeRegs(op1);
 
-        assert(op1->OperGet() == GT_CALL);
-        assert((op1->gtFlags & GTF_SPILLED) == 0);
-    }
+        GenTreeCall* call = op1->AsCall();
+        ReturnTypeDesc* retTypeDesc = call->GetReturnTypeDesc();
+        unsigned regCount = retTypeDesc->GetReturnRegCount();
+        assert(regCount == MAX_RET_REG_COUNT);
 
+        // Handle circular dependency between call allocated regs and ABI return regs.
+        //
+        // It is possible under LSRA stress that originally allocated regs of call node,
+        // say rax and rdx, are spilled and reloaded to rdx and rax respectively.  But
+        // GT_RETURN needs to  move values as follows: rdx->rax, rax->rdx. Similar kind
+        // kind of circular dependency could arise between xmm0 and xmm1 return regs.
+        // Codegen is expected to handle such circular dependency.
+        //
```

Comment:
It would be great if we could ensure that we have a test case / stress mode combination that will cause this to happen.